### PR TITLE
Assert correct elements are removed from hashes larger than limit.

### PR DESCRIPTION
**Pull request by ptolts**
Adds missing test coverage for Tracestates larger than 32.

### CODE REVIEW

**Review by fbogsany — CHANGES_REQUESTED** [review]
This is technically incorrect. From https://github.com/open-telemetry/opentelemetry-ruby/pull/1533#issuecomment-1783535817

> Followup: the [W3C `trace-context` spec](https://www.w3.org/TR/trace-context/#tracestate-limits) states:
> 
> > In a situation where `tracestate` needs to be truncated due to size limitations, the vendor _MUST_ truncate whole entries. Entries larger than 128 characters long _SHOULD_ be removed first. Then entries _SHOULD_ be removed starting from the end of `tracestate`. Note that other truncation strategies like safe list entries, blocked list entries, or size-based truncation _MAY_ be used, but are highly discouraged. Those strategies decrease the interoperability of various tracing vendors.

Our implementation is wrong and we should remove entries from the end rather than the start (and, of course, we should remove entries > 128 characters first).

**Comment by ptolts** [comment]
I'll close, thanks Bogs.

**Comment by fbogsany** [comment]
@ptolts could you please click the link in https://github.com/open-telemetry/opentelemetry-ruby/pull/1535#issuecomment-1783522860 to authorize the CLA?

File: api/test/opentelemetry/trace/tracestate_test.rb [hunk 1]
```diff
@@ -74,9 +74,12 @@
...
     end
-    it 'will not exceed 32 members' do
-      tracestate = OpenTelemetry::Trace::Tracestate.from_hash(33.times.collect { |n| [n.to_s, n.to_s] }.to_h)
-      _(tracestate.to_h.size).must_be :<=, 32
+    it 'will not exceed 32 members and drops excess from the beginning' do
+      input_hash = 33.times.collect { [('a'..'z').to_a.sample(5).join, ('a'..'z').to_a.sample(5).join] }.to_h
+      tracestate = OpenTelemetry::Trace::Tracestate.from_hash(input_hash)
+      output_hash = tracestate.to_h
+      _(output_hash.size).must_be :<=, 32
+      _(output_hash).must_equal(input_hash.drop(1).to_h)
```
Comment (fbogsany) [inline api/test/opentelemetry/trace/tracestate_test.rb:82]:
```suggestion
      _(output_hash).must_equal(input_hash.first(32).to_h)
```

**Comment by kaylareopelle** [comment]
Hi, @ptolts! Checking in on this PR. Would you be open to signing the CLA?

**Pull request closed without merging**